thetis/recovery.py

```diff
@@ -0,0 +1,125 @@
+from .utility import *
+
+
+class VorticityCalculator2D(object):
+    r"""
+    Linear solver for recovering fluid vorticity.
+
+    It is recommended that the vorticity is sought
+    in :math:`\mathbb P1` space.
+
+    :arg vorticity_2d: :class:`Function` to hold calculated vorticity.
+    :arg solver_obj: :class:`FlowSolver2d` object.
+    :kwargs: to be passed to the :class:`LinearVariationalSolver`.
+    """
+    @PETSc.Log.EventDecorator("thetis.VorticityCalculator2D.__init__")
+    def __init__(self, vorticity_2d, solver_obj, **kwargs):
+        fs = vorticity_2d.function_space()
+        dim = fs.mesh().topological_dimension()
+        if dim != 2:
+            raise ValueError(f'Dimension {dim} not supported')
+        if element_continuity(fs.ufl_element()).horizontal != 'cg':
+            raise NotImplementedError
+        n = FacetNormal(fs.mesh())
+
+        # Weak formulation
+        test = TestFunction(fs)
+        uv, elev = split(solver_obj.fields.solution_2d)
+        a = TrialFunction(fs)*test*dx
+        L = -inner(perp(uv), grad(test))*dx \
+            + dot(perp(uv), n)*test*ds
+
+        # Setup vorticity solver
+        prob = LinearVariationalProblem(a, L, vorticity_2d)
+        kwargs.setdefault('solver_parameters', {
+            "ksp_type": "cg",
+            "pc_type": "bjacobi",
+            "sub_pc_type": "ilu",
+        })
+        self.solver = LinearVariationalSolver(prob, **kwargs)
+
+    @PETSc.Log.EventDecorator("thetis.VorticityCalculator2D.solve")
+    def solve(self):
+        self.solver.solve()
+
+
+class HessianRecoverer2D(object):
+    r"""
+    Linear solver for recovering Hessians.
+
+    Hessians are recoved using double :math:`L^2`
+    projection, which is implemented using a
+    mixed finite element method.
+
+    :arg hessian_2d: :class:`Function` to hold recovered Hessian.
+    :arg solver_obj: :class:`FlowSolver2d` object.
+    :kwargs: to be passed to the :class:`LinearVariationalSolver`.
+    """
+    @PETSc.Log.EventDecorator("thetis.HessianRecoverer2D.__init__")
+    def __init__(self, hessian_2d, solver_obj, field, **kwargs):
```

Comment:
I would prefer that these routines do not depend on `solver_obj` (applies to VorticityCalculator2D as well). After all, you only use it to access `solver_obj.fields[field]` so you could just pass in the field itself. This implies a little more boilerplate code by the user but you already need some, e.g. to define the TensorFunctionSpace. If this is too much we could add a helper function somewhere.

Reply:
Sure, I can do that.